Check user exists for setup redirect

escherize opened the pull request:
This change fixes #12874. 

When setting `setup-token` via the `MB_SETUP_TOKEN` environment variable, the setting cannot be changed. We were trying to clear the setting, and use that value on the frontend to skip the setup flow.

This change introduces a new setting: `has-user-setup` which is true iff one or more users have been registered on the instance. `has-user-setup` is used by the frontend to determine whether to redirect to the setup flow, or not.

This setting is also used to allow only _one_ user to be added via a post to `/api/setup`, as the user in the issue was doing.

also note, that tests use the `/api/setup` endpoint to create multiple users, so I have added a check to `config/is-test?` that will allow that behavior.

escherize committed:
Add has-user-setup setting

- iff no user is setup aka has-user-setup == false, then redirect to /setup
- when set from the env variable the setup-token is completely
  immutable, so can remove clear-token!

escherize committed:
make /api/setup only setup accounts when not has-user-setup

- react router changes to respect 'has-user-setup' setting

escherize committed:
Merge branch 'master' of github.com:metabase/metabase into check-user-exists-for-setup-redirect

escherize committed:
include docstring for print-setup-url

escherize committed:
allow tests to setup multiple users via POST /api/setup

escherize committed:
remove unused function hasSetupToken

escherize committed:
Merge branch 'master' of github.com:metabase/metabase into check-user-exists-for-setup-redirect

camsaul commented on src/metabase/core.clj:
```diff
@@ -60,6 +59,12 @@
                                    setup-url
                                    "\n\n")))))
 
+(defn- -init-create-setup-token
```
I know you didn't come up with the name `-init-create-setup-token` here but since you're touching it I noticed it. I think we should change the name to fit our coding conventions elsewhere. I think something like `create-setup-token-and-log-setup-url!` would be a 100x better name for this function.

1. We don't really prefix anything else with a hyphen no matter how secret/internal it is
2. We suffix DML type things with `!` (https://guide.clojure.style/#naming-unsafe-functions)
3. `-init-create-setup-token` doesn't really tell you much about what this function does

camsaul commented on src/metabase/core.clj:
```diff
@@ -44,10 +44,9 @@
 
-(defn- -init-create-setup-token
-  "Create and set a new setup token and log it."
+(defn print-setup-url
```
You should make this private https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

The main reason is so if we want to refactor it in the future we don't have to look around the codebase and make sure it's not used anywhere else. (Also when you look thru this namespace to use something from it the surface area of it is a little smaller and it's clearer which things are mean to be used elsewhere)

escherize replied:
Good catch, thanks!

camsaul commented on src/metabase/api/setup.clj:
```diff
@@ -39,6 +39,9 @@
 (defn- setup-create-user! [{:keys [email first-name last-name password]}]
+  (when-not config/is-test?
+    ;; many tests use /api/setup to setup multiple users
+    (assert (not (setup/has-user-setup)) "setup-create-user! can only be called when no users have been setup."))
```
1. This should be i18n'ed since there's a chance it can be hit by someone who isn't a actively hacking on Metabase
2. This should throw an `ExceptionInfo` with a `:status-code`. Otherwise I think this would probably end up returning a 500 which is not appropriate here
```suggestion
    (when-not (setup/has-user-setup)
      (throw (ex-info (tru "Setup endpoints can not be used after a user has been created.") {:status-code 400})))
```

camsaul replied:
I think having that `config/is-test?` flag is actually not the best way to go about this since it effectively makes this logic impossible to test. I think adding a dynamic variable like `*disallow-setup-api-calls-after-first-user-is-created*` (default `false`) and then binding that to `true` in relevant tests is better than just a blanket disabling of this logic in tests

escherize replied:
Makes sense. I have added `*allow-setup-api-calls-after-first-user-is-created*` and am working on testing it

camsaul commented on src/metabase/setup.clj:
```diff
@@ -25,12 +26,14 @@
-  "Clear the setup token if it exists and reset it to `nil`."
-  []
-  (setting/set-value-of-type! :string :setup-token nil))
+(defsetting has-user-setup
+  "A value that is true iff the metabase instance has one or more users registered."
+  :visibility :public
+  :type       :boolean
+  :setter     :none
+  :getter     (fn [] (db/exists? User)))
```
We can't ever become un-set-up after we're set up (we specifically disallow deleting the last User) and since this is a `:public` Setting that gets returned every time someone loads up the app I think it's worth optimizing out the application DB hit. One little hit like this is not a huge deal but they tend to add up over time if we don't try to be aggressive about eliminating them where we can.

I would do something like this:

```suggestion
(defsetting has-user-setup
  "A value that is true iff the metabase instance has one or more users registered."
  :visibility :public
  :type       :boolean
  :setter     :none
  ;; Once a User is created it's impossible for this to ever become falsey -- deleting the last User is disallowed.
  ;; After this returns true once the result is cached and it will continue to return true forever without any
  ;; additional DB hits.
  :getter     (let [user-exists? (atom false)]
                (fn []
                  (or @user-exists?
                      (reset! user-exists? (db/exists? User))))))
```

dpsutton replied:
I'm still a bit hazy on how the caching system works but would the cache not prevent this from hitting the db all the time?

camsaul replied:
Caching is part of the default get-value-from-application-DB getters but for a Setting with a completely custom getter the cache is bypassed. Calling `(has-user-setup)` or `(setting/get :has-user-setup)` is effectively the same as just calling the `:getter` function here.

This is intentional, since custom getters are usually returning calculated values caching them wouldn't be desirable in most cases. For example because the cache is set up to flush when rows in the `setting` Table change in the app DB, it wouldn't pick up changes to the `User` table and could end up returning the wrong value if this was cached

dpsutton replied:
Makes sense. Thanks

camsaul reviewed (dismissed):
This looks like it's on the right track to me but it **really** needs some tests. Having no test coverage for code in absolutely critical parts of the application like the setup is just too risky (especially since being able to hit setup a second time is a huge security risk -- especially now that we're not clearing out the token anymore). Please add some tests to verify that the endpoints in question actually return a 400 response if a User already exists. Ping me again when tests are added and I will 👍

escherize committed:
Merge branch 'master' of github.com:metabase/metabase into check-user-exists-for-setup-redirect

escherize committed:
Adds a test for single login only

- updates tests in api.setup-test to work with
- we now throw an ex-info with status-code 403 when has-user-setup and
  /api/setup route is hit (with error message mentioning the route)
- more code review responses

escherize committed:
Add missing docstring

escherize committed:
bump tests

- fixup test to not delete users, (and set locale to spanish...)

dpsutton commented on src/metabase/api/setup.clj:
```diff
@@ -38,21 +38,33 @@
   (su/with-api-error-message (s/constrained su/NonBlankString setup/token-match?)
     "Token does not match the setup token."))
 
+(def ^:dynamic ^:private *disallow-api-setup-after-first-user-is-created*
```
this should be `*allow-api-setup-after-first-user-created* false` and then the test can change this to `true` as needed.

dpsutton commented on src/metabase/api/setup.clj:
```diff
@@ -38,21 +38,33 @@
+  true)
+
 (defn- setup-create-user! [{:keys [email first-name last-name password]}]
+  (when (and *disallow-api-setup-after-first-user-is-created* (setup/has-user-setup))
```
the logic here seems weird. It doesn't feel like an override here. I know we can deMorgan our way into this but the logic could be more clear.
`(when (and (setup/has-user-setup) (not *allow-api-setup...*)` reads more like what my brain interprets it as. There's already a user setup and the override is not in place. It feels backwards to check the override before even the condition. I know its equivalent but my brain is struggling for some reason.

escherize replied:
Good point, its easy from my pov to focus on the override itself. But it makes more sense to highlight on the triggering condition (aka there is a user already).

dpsutton commented on src/metabase/core.clj:
```diff
@@ -44,10 +44,9 @@
-(defn- -init-create-setup-token
-  "Create and set a new setup token and log it."
+(defn- print-setup-url
+  "Used to print the setup url during instance initialization."
```
"Used to" adds nothing here. "Print the setup url during instance initialization" is the same but without being verbose.

dpsutton commented on test/metabase/api/setup_test.clj:
```diff
@@ -272,6 +274,39 @@
+                           :email      (mt/random-email)
+                           :password   "p@ssword1"}}]
+
+        (is (setup/has-user-setup))
```
If this test is run with `clj -X:dev:test :only metabase.api.setup-test/create-superuser-only-once-test` i suspect this line will fail? It seems we will need to ensure that the user information is completely mocked in this test

escherize replied:
There are actually 4 users setup in the test database, so has-user-setup is 'set'. I was working on a function wrapper that would save, and delete them, run the test, and reinstate them, but cannot get it to work in a way that doesn't pollute the test db and break _other_ tests.

dpsutton replied:
That's fair. I'm just not positive that it is always the case that there are four users in the database under scenarios where you run the tests for an individual namespace from the cli.

escherize replied:
Gotcha, I'll remove that line and add a test that takes that into account.

dpsutton commented on test/metabase/api/setup_test.clj:
```diff
@@ -272,6 +274,39 @@
+        (with-redefs [setup/has-user-setup (let [value (atom false)]
+                                             (fn
+                                               ([] @value)
+                                               ([x] (reset! value x))))]
```
the setting has a `:setter :none`. Should this throw rather than reset the value? or is this just a helper to change the value in the test?

escherize replied:
I'll simplify this. As you said it doesn't need to be settable, all it needs to do is return a value when called with 0 args, and not accept 1 arg. Would it be worth using an actual setting with a :getter that returns false?

dpsutton replied:
I guess I'm wondering if this should be just `(binding [setup/has-user-setup (constantly false))`. Or if you want to change it after a few tests,

`(let [setup? (atom false)] (binding [setup/has-user-setup (fn [] @setup?)] (test) (reset! setup? true) (test asserting that you can do something now)`

escherize replied:
I thought about the `(constantly false)` piece a bit: I think it shouldn't be because I want an exception to occur if `setup/has-user-setup` arity-1 version gets called, because thats how the setting behaves with a `:none` `:getter`.

escherize replied:
The latest PR includes the switching to `true` of `setup/has-user-setup`.

dpsutton commented on test/metabase/api/setup_test.clj:
```diff
@@ -272,6 +274,39 @@
+          (mt/discard-setting-changes [site-name
+                                       site-locale ;; anon-tracking-enabled
+                                       admin-email]
+            (fn []
```
`mt/discard-setting-changes` is a macro that thunks its body so i don't think you want the manual `(fn [] ...)` bits here

escherize replied:
How embarrassing! 😆

dpsutton commented on src/metabase/api/setup.clj:
```diff
@@ -38,21 +38,33 @@
+    ;; many tests use /api/setup to setup multiple users, so *first-user-only* is redefined by them
+    (throw (ex-info
+            (tru "The /api/setup route can only be used to create the first user, however a user currently exists.")
+            {:status-code 403
```
I'm wondering if we should just return the 403 here and not the other information.

escherize replied:
I noticed some other places ex-info used return information, but perhaps the `has-users-setup` value is too internal to be useful here.

escherize committed:
addressing code review concerns

- disallow -> allow
- has-user-settings test: do not assume the test db has users in it
- update test

dpsutton commented on src/metabase/models/setting.clj:
```diff
@@ -932,7 +932,7 @@
     (into
      {}
-     (comp (filter (fn [[_ setting]]
+     (comp (filter (fn [[_setting-name setting]]
```
just an aside, I love the convention of `_setting-name` to make this easier to read. Thanks

dpsutton commented:
Works for me. I grabbed the jar from here,

```shell
MB_SETUP_TOKEN=0e64bb35-187a-458b-a85d-c182398b238a \
  MB_DB_CONNECTION_URI="postgres://<redact>/setup1" \
  MB_JETTY_PORT=3006 java -jar metabase.jar
```

and in another terminal

```shell
❯ http POST http://localhost:3006/api/setup token=0e64bb35-187a-458b-a85d-c182398b238a "user[first_name]"=dan "user[last_name]"=sutton "user[email]"="dan@metabase.com" "user[password]"=bugbash1 "prefs[site_name]"=test
HTTP/1.1 200 OK
 [stuff omitted]

{
    "id": "65fef68a-e73c-4d10-8c54-e837817a6ccf"
}

❯ http POST http://localhost:3006/api/setup token=0e64bb35-187a-458b-a85d-c182398b238a "user[first_name]"=dan "user[last_name]"=sutton "user[email]"="dan1@metabase.com" "user[password]"=bugbash1 "prefs[site_name]"=test
HTTP/1.1 403 Forbidden

The /api/setup route can only be used to create the first user, however a user currently exists.
```
and of course

```sql
setup1=# select count(*) from core_user;
 count
-------
     1
(1 row)
```

And the app took me to the login screen instead of the setup flow.

dpsutton commented on test/metabase/api/setup_test.clj:
```diff
@@ -272,6 +274,39 @@
+            (http/client :post 200 "setup" body))
+
+          ;; In the non-test context, this is 'set' iff there is one or more users, and doesn't have to be toggled
+          (setup/has-user-setup true)
```
I think i'd rather see `(reset! allow-setup true)` so no one gets the impression that this setting is settable, even though we are `redef`ing it above. I think I'd also like some user count assertions. Get the count of users in the let, assert it increments after you 200 post, then assert it does not change again after you 403 post.

dpsutton replied:
The user count is what is actually super important here and what we want to assert.

camsaul commented:
I'm 👍 on this WRT to the changes I requested but I'll defer to @dpsutton to give the final 👍 since he has more context
